### PR TITLE
Fix NFD computation

### PR DESCRIPTION
**Pull request by JoanFM**
I am trying to change the NFD computation according to https://unicode.org/reports/tr15/#Description_Norm

Thee changes are:

- Get the `range` from the `nfd_map` and apply the `decomposition` recursively. (Edit) According to @iamlemec, the `nfd_map` is constructed in a way that the recursion is applied
- Sort results according to the `Canonical_Combining_Class`.

***TODO***

- [x] Properly fill the `unicode_canonical_class` map
- [x] Do some tests and validate the implementation
- [x] Fix the implementation avoiding potential problem with inserting into a vector while iterating over it

### CODE REVIEW

File: examples/server/tests/features/embeddings.feature [hunk 1]
```diff
@@ -16,12 +16,12 @@ Feature: llama.cpp server
     Then  the server is starting
     Then  the server is healthy
 
-  Scenario: Embedding
-    When embeddings are computed for:
+  Scenario: Tokenize / Detokenize complex
```
Comment (JoanFM) [inline examples/server/tests/features/embeddings.feature:19]: this should be added ,not substituted

**Comment by ggerganov** [comment]
Using your example, I added the following test to `convert-hf-to-gguf-update.py`:

```diff
@@ -252,6 +255,7 @@ tests = [
     "3333333",
     "33333333",
     "333333333",
+    "乡—서울에서",
     chktxt,
 ]
```

The unit test using the bert vocab still fails with this PR:

```bash
# regenerate test data
python3 convert-hf-to-gguf-update.py <hf_token>

# run tests
make -j tests && ./tests/test-tokenizer-0 ./models/ggml-vocab-bert-bge.gguf
```

Shouldn't it pass with the NFD fixes?

**Comment by JoanFM** [comment]
> Using your example, I added the following test to `convert-hf-to-gguf-update.py`:
> 
> ```diff
> @@ -252,6 +255,7 @@ tests = [
>      "3333333",
>      "33333333",
>      "333333333",
> +    "乡—서울에서",
>      chktxt,
>  ]
> ```
> 
> The unit test using the bert vocab still fails with this PR:
> 
> ```shell
> # regenerate test data
> python3 convert-hf-to-gguf-update.py <hf_token>
> 
> # run tests
> make -j tests && ./tests/test-tokenizer-0 ./models/ggml-vocab-bert-bge.gguf
> ```
> 
> Shouldn't it pass with the NFD fixes?

Hey @ggerganov ,

Can you tell me the exact steps to reproduce?

**Comment by JoanFM** [comment]
> ```shell
> ggml-vocab-bert-bge.gguf
> ```

What exactly I need to pass to the `convert-hf-to-gguf-update.py` call?

**Comment by ggerganov** [comment]
You need a HuggingFace account: https://huggingface.co/settings/tokens

Pass the read token from the link above

If you don't have HF account, I can push the test to your branch manually?

**Comment by JoanFM** [comment]
> You need a HuggingFace account: https://huggingface.co/settings/tokens
> 
> Pass the read token from the link above
> 
> If you don't have HF account, I can push the test to your branch manually?

u can feel free to push yes

**Comment by JoanFM** [comment]
> You need a HuggingFace account: https://huggingface.co/settings/tokens
> 
> Pass the read token from the link above
> 
> If you don't have HF account, I can push the test to your branch manually?

I get the error:

```INFO:convert-hf-to-gguf-update:File models/tokenizers/command-r/tokenizer_config.json downloaded successfully
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/joan/workspace/ollama/llm/llama.cpp/convert-hf-to-gguf-update.py", line 139, in <module>
    tokenizer = AutoTokenizer.from_pretrained(f"models/tokenizers/{name}")
  File "/home/joan/workspace/ollama/llm/llama.cpp/gguf-research/venv/lib/python3.10/site-packages/transformers/models/auto/tokenization_auto.py", line 819, in from_pretrained
    config = AutoConfig.from_pretrained(
  File "/home/joan/workspace/ollama/llm/llama.cpp/gguf-research/venv/lib/python3.10/site-packages/transformers/models/auto/configuration_auto.py", line 928, in from_pretrained
    config_dict, unused_kwargs = PretrainedConfig.get_config_dict(pretrained_model_name_or_path, **kwargs)
  File "/home/joan/workspace/ollama/llm/llama.cpp/gguf-research/venv/lib/python3.10/site-packages/transformers/configuration_utils.py", line 631, in get_config_dict
    config_dict, kwargs = cls._get_config_dict(pretrained_model_name_or_path, **kwargs)
  File "/home/joan/workspace/ollama/llm/llama.cpp/gguf-research/venv/lib/python3.10/site-packages/transformers/configuration_utils.py", line 686, in _get_config_dict
    resolved_config_file = cached_file(
  File "/home/joan/workspace/ollama/llm/llama.cpp/gguf-research/venv/lib/python3.10/site-packages/transformers/utils/hub.py", line 369, in cached_file
    raise EnvironmentError(
OSError: models/tokenizers/llama-bpe does not appear to have a file named config.json. Checkout 'https://huggingface.co/models/tokenizers/llama-bpe/tree/None' for available files.
```

**Comment by ggerganov** [comment]
Pull and run:

```bash
make -j tests && ./tests/test-tokenizer-0 ./models/ggml-vocab-bert-bge.gguf
```

This compares the resulting tokens when using `hf/tokenizers` and `llama.cpp`. The test that fails is with the string `乡—서울에서`:

- HF tokenizers produces the tokens: `[100, 1517, 1461, 30008, 29999, 30014, 30022, 29999, 30009, 29997, 30008]`
- `llama.cpp` produces the tokens: `[100 1517 100]`

**Comment by JoanFM** [comment]
> Pull and run:
> 
> ```shell
> make -j tests && ./tests/test-tokenizer-0 ./models/ggml-vocab-bert-bge.gguf
> ```
> 
> This compares the resulting tokens when using `hf/tokenizers` and `llama.cpp`. The test that fails is with the string `乡—서울에서`:
> 
> * HF tokenizers produces the tokens: `[100, 1517, 1461, 30008, 29999, 30014, 30022, 29999, 30009, 29997, 30008]`
> * `llama.cpp` produces the tokens: `[100, 1517, 100]`

How can I know the `hf` model or tokenizer that is using? Like this I can debug better

**Comment by ggerganov** [comment]
The reference HF tokenizer model is listed in the `convert-hf-to-gguf-update.py` script:

https://github.com/ggerganov/llama.cpp/blob/acdce3cdef6fc2f0b7b5623231fd7762c0884d1c/convert-hf-to-gguf-update.py#L56-L71

For the BERT model "bert-bge", the link is: https://huggingface.co/BAAI/bge-small-en-v1.5/tree/main

You can clone that repository and write a Python script that imports `AutoTokenizer` from the `transformers` package and loads the tokenizer model from that repository. This is the reference tokenization that we try to match in `llama.cpp`

The `convert-hf-to-gguf-update.py` scripts already does that and generates a set of unit tests

**Comment by JoanFM** [comment]
> bge

okey, thanks I will investigate further thx

**Comment by ggerganov** [comment]
Yes, it's possible. I was trying to add a test for the NFD changes. Can you provide a text that does not work on `master` but works with this PR?

**Comment by JoanFM** [comment]
> Yes, it's possible. I was trying to add a test for the NFD changes. Can you provide a text that does not work on `master` but works with this PR?

I am trying but it seems quite hard,

I find cases where the NFD output differs, but the end tokenization results in the same tokens.

```The façade of the café was uniquely designed with naïve motifs and coöperate signage.```

This is expanded ny NFD but the extra codepoints do not reflect in the tokenization. I have tried with mutliple examples and no success.

If you want I can close this PR. Now I have a little bit more clarity and most of the cases I would just ignore NFC and NFD for my problem and hope for it to work most of the time.

**Comment by JoanFM** [comment]
> > Yes, it's possible. I was trying to add a test for the NFD changes. Can you provide a text that does not work on `master` but works with this PR?
> 
> I am trying but it seems quite hard,
> 
> I find cases where the NFD output differs, but the end tokenization results in the same tokens.
> 
> `The façade of the café was uniquely designed with naïve motifs and coöperate signage.`
> 
> This is expanded ny NFD but the extra codepoints do not reflect in the tokenization. I have tried with mutliple examples and no success.
> 
> If you want I can close this PR. Now I have a little bit more clarity and most of the cases I would just ignore NFC and NFD for my problem and hope for it to work most of the time.

I will try to test with some emojis like "👨‍👩‍👧‍👦"

**Comment by ggerganov** [comment]
If you think the `find` -> `equal_range` makes sense, we can merge. But ideally it would be nice to have a test that validates the behaviour

**Comment by JoanFM** [comment]
> If you think the `find` -> `equal_range` makes sense, we can merge. But ideally it would be nice to have a test that validates the behaviour

Seems to make sense in the strict sense of the algorithm, but I tried several cases quite complex and could not find the case, because it seems that the second part is not part of the vocabulary.

**Comment by JoanFM** [comment]
> > If you think the `find` -> `equal_range` makes sense, we can merge. But ideally it would be nice to have a test that validates the behaviour
> 
> Seems to make sense in the strict sense of the algorithm, but I tried several cases quite complex and could not find the case, because it seems that the second part is not part of the vocabulary.

Maybe @iamlemec can provide more context and finds a case where it could change?

**Comment by iamlemec** [comment]
Ok, so I've searched programmatically over the `UnicodeData.txt` data and it appears there are no codepoints that decompose into more than one letter. Ligatures like "æ" do not get decomposed here. And since the letter is always listed first, I'm not sure there exists such an example. To me, this points in the direction of keeping `find` as is and possibly turning `unicode_map_nfd` into a regular map.

Now, this business with the Korean (Hangul) letters is interesting as a separate issue. Looks like the HF tokenizer handles those programmatically, as recommended. Since there are `19 x 21 x 28` character combinations, you can just invert the codepoint like a flat index. I would be happy to address this in another PR if that works for folks.

**Comment by JoanFM** [comment]
Let's close this, thanks for the help.

I got some nice insights into this new world of Unicode and Normalization.

Thanks @ggerganov , @iamlemec and @teleprint-me !

**Pull request closed without merging**